Mathlib/Geometry/Manifold/LocalDiffeomorph.lean

```diff
@@ -0,0 +1,272 @@
+def toPartialHomeomorph : PartialHomeomorph M N where
+  toPartialEquiv := Φ.toPartialEquiv
+  open_source := Φ.open_source
+  open_target := Φ.open_target
+  continuousOn_toFun := Φ.contMDiffOn_toFun.continuousOn
+  continuousOn_invFun := Φ.contMDiffOn_invFun.continuousOn
+
+/-- The inverse of a local diffeomorphism. -/
+protected def symm : PartialDiffeomorph J I N M n where
+  toPartialEquiv := Φ.toPartialEquiv.symm
+  open_source := Φ.open_target
+  open_target := Φ.open_source
+  contMDiffOn_toFun := Φ.contMDiffOn_invFun
+  contMDiffOn_invFun := Φ.contMDiffOn_toFun
+
+protected theorem contMDiffOn : ContMDiffOn I J n Φ Φ.source :=
+  Φ.contMDiffOn_toFun
+
+protected theorem mdifferentiableOn : MDifferentiableOn I J Φ Φ.source :=
+  (Φ.contMDiffOn).mdifferentiableOn hn
+
+protected theorem mdifferentiableAt {x : M} (hx : x ∈ Φ.source) : MDifferentiableAt I J Φ x :=
+  (Φ.mdifferentiableOn hn x hx).mdifferentiableAt (Φ.open_source.mem_nhds hx)
+
+/- We could add lots of additional API (following `Diffeomorph` and `PartialHomeomorph`), such as
+* further continuity and differentiability lemmas
+* refl and trans instances; lemmas between them.
+As this declaration is meant for internal use only, we keep it simple. -/
+end PartialDiffeomorph
+end PartialDiffeomorph
+
+variable {M N}
+
+/-- `f : M → N` is called a **`C^n` local diffeomorphism at *x*** iff there exist
+  open sets `U ∋ x` and `V ∋ f x` and a diffeomorphism `Φ : U → V` such that `f = Φ` on `U`. -/
+def IsLocalDiffeomorphAt (f : M → N) (x : M) : Prop :=
+  ∃ Φ : PartialDiffeomorph I J M N n, x ∈ Φ.source ∧ EqOn f Φ Φ.source
+
+/-- `f : M → N` is called a **`C^n` local diffeomorphism on *s*** iff it is a local diffeomorphism
+  at each `x : s`. -/
+def IsLocalDiffeomorphOn (f : M → N) (s : Set M) : Prop :=
+  ∀ x : s, IsLocalDiffeomorphAt I J n f x
+
+/-- `f : M → N` is a **`C^n` local diffeomorphism** iff it is a local diffeomorphism
+at each `x ∈ M`. -/
+def IsLocalDiffeomorph (f : M → N) : Prop :=
+  ∀ x : M, IsLocalDiffeomorphAt I J n f x
+
+lemma isLocalDiffeomorphOn_iff {f : M → N} (s : Set M) :
+    IsLocalDiffeomorphOn I J n f s ↔ ∀ x : s, IsLocalDiffeomorphAt I J n f x := by rfl
+
+lemma isLocalDiffeomorph_iff {f : M → N} :
+    IsLocalDiffeomorph I J n f ↔ ∀ x : M, IsLocalDiffeomorphAt I J n f x := by rfl
+
+theorem isLocalDiffeomorph_iff_isLocalDiffeomorphOn_univ {f : M → N} :
+    IsLocalDiffeomorph I J n f ↔ IsLocalDiffeomorphOn I J n f Set.univ :=
+  ⟨fun hf x ↦ hf x, fun hf x ↦ hf ⟨x, trivial⟩⟩
+
+lemma IsLocalDiffeomorph.isLocalDiffeomorphOn {f : M → N} (s : Set M)
+    (hf : IsLocalDiffeomorph I J n f) : IsLocalDiffeomorphOn I J n f s := fun x ↦ hf x
+
+variable {n}
+
+/-! # Basic properties of local diffeomorphisms -/
+section Basic
+variable {f : M → N} {s : Set M} {x : M}
+
+/-- A `C^n` local diffeomorphism at `x` is `C^n` differentiable at `x`. -/
+lemma contMDiffAt_of_isLocalDiffeomorphAt (hf : IsLocalDiffeomorphAt I J n f x) :
```

Comment:
This lemma should be named `IsLocalDiffeomorphAt.contMDiffAt`, to enable dot notation. Same thing for most lemmas below.